pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_shoot.go

```diff
@@ -0,0 +1,135 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graph
+
+import (
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	toolscache "k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+)
+
+func (g *graph) setupShootWatch(informer cache.Informer) {
+	informer.AddEventHandler(toolscache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			shoot, ok := obj.(*gardencorev1beta1.Shoot)
+			if !ok {
+				return
+			}
+			g.handleShootCreateOrUpdate(shoot)
+		},
+
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldShoot, ok := oldObj.(*gardencorev1beta1.Shoot)
+			if !ok {
+				return
+			}
+
+			newShoot, ok := newObj.(*gardencorev1beta1.Shoot)
+			if !ok {
+				return
+			}
+
+			if !apiequality.Semantic.DeepEqual(oldShoot.Spec.SeedName, newShoot.Spec.SeedName) ||
+				!apiequality.Semantic.DeepEqual(oldShoot.Status.SeedName, newShoot.Status.SeedName) ||
+				!apiequality.Semantic.DeepEqual(oldShoot.Spec.SecretBindingName, newShoot.Spec.SecretBindingName) ||
+				!apiequality.Semantic.DeepEqual(oldShoot.Spec.CloudProfileName, newShoot.Spec.CloudProfileName) ||
+				!gardencorev1beta1helper.ShootAuditPolicyConfigMapRefEqual(oldShoot.Spec.Kubernetes.KubeAPIServer, newShoot.Spec.Kubernetes.KubeAPIServer) ||
+				!gardencorev1beta1helper.ShootDNSProviderSecretNamesEqual(oldShoot.Spec.DNS, newShoot.Spec.DNS) ||
+				!gardencorev1beta1helper.ShootSecretResourceReferencesEqual(oldShoot.Spec.Resources, newShoot.Spec.Resources) {
+				g.handleShootCreateOrUpdate(newShoot)
+			}
+		},
+
+		DeleteFunc: func(obj interface{}) {
+			if tombstone, ok := obj.(toolscache.DeletedFinalStateUnknown); ok {
+				obj = tombstone.Obj
+			}
+			shoot, ok := obj.(*gardencorev1beta1.Shoot)
+			if !ok {
+				return
+			}
+			g.handleShootDelete(shoot)
+		},
+	})
+}
+
+func (g *graph) handleShootCreateOrUpdate(shoot *gardencorev1beta1.Shoot) {
+	g.lock.Lock()
+	defer g.lock.Unlock()
+
+	g.deleteAllIncomingEdges(VertexTypeCloudProfile, VertexTypeShoot, shoot.Namespace, shoot.Name)
+	g.deleteAllIncomingEdges(VertexTypeConfigMap, VertexTypeShoot, shoot.Namespace, shoot.Name)
+	g.deleteAllIncomingEdges(VertexTypeNamespace, VertexTypeShoot, shoot.Namespace, shoot.Name)
+	g.deleteAllIncomingEdges(VertexTypeSecret, VertexTypeShoot, shoot.Namespace, shoot.Name)
+	g.deleteAllIncomingEdges(VertexTypeSecretBinding, VertexTypeShoot, shoot.Namespace, shoot.Name)
+	g.deleteAllOutgoingEdges(VertexTypeShoot, shoot.Namespace, shoot.Name, VertexTypeSeed)
```

Comment:
I was wondering if we can generalize this so that whenever a new reference type is added in the future, you don't have to consider it here. IIUC, it's always delete all incoming and outgoing edges, isn't it?

Reply:
> IIUC, it's always delete all incoming and outgoing edges, isn't it?

Yes, exactly, all edges that it could have created in earlier invocations (i.e., all edges that are created further down in this function). Do you have a suggestion for a potential generalization?

Reply:
As discussed, this can't be done in some cases (BackupBucket) as it would remove irrecoverable information.